Documentation/gopen.1
.\"gopen(1) man page
.\"put together by Martin Brecher <martin@gnustep.de>
.\"Copyright (C) 2005 Free Software Foundation, Inc.
.\"
.\"Process this file with
.\"groff -man -Tascii gopen.1
.\"
.TH GOPEN 1 "August 2003" GNUstep "GNUstep System Manual"
.SH NAME
gopen \- open files
.SH SYNOPSIS
.B gopen
.RB [ -a 
.IR application ]
.RB [ -o ]
.RB [ -p ]
.RB [ -NXHost 
.IR hostname ]
.RI [ filename ]
.P
.SH DESCRIPTION
The 
.B gopen
command allows you open a file (or directory) as if you had double clicked
the object's icon.
.PP
It is possible to specify one or more 
.I filenames
which are interpreted relative to the current working directory. 
.PP
By default, 
.B gopen
will open
.I filename
with the application currently assigned to the file's extension. But by
specifing the 
.B -a
flag on the command line you can tell
.B gopen
to open the file with anther application.
.PP
You can also directly print a file without (
.B -p
) or with (
.B -p -o 
) opening it.
.PP
As with most other GNUstep software, it is easily possible
to attach the process to a remote window server using the
.B -NXHost
flag.
.PP
.SH OPTIONS
.IP "\fB-a \fIapplication"
use
.I application
to open
.I filename
.IP "\fB-o"
open 
.I filename
(may be used in conjunction with
.B -p
).
.IP "\fB-p"
print
.I filename
instead of opening.
.IP "\fB-NXHost \fIhostname"
attach to remote window server on
.I hostname
.PP
.SH EXAMPLES
Open all files with a .txt extension in the current directory:
.PP
.I gopen *.txt
.PP
To open a source file in CodeEditor (instead of the current default 
application), type:
.PP
.I gopen -a CodeEditor MySourceFile.m
.PP
To print the .plan file in your home folder, use the
.B -p
flag:
.PP
.I gopen -p ~/.plan
.PP
.SH BUGS
Using the
.B -a
.I application
argument,
.B gopen
allows you to open any kind of file with any application you want. This does,
however, not work with applications employing the NSDocument architecture
as they will only receive filenames whose types the application has registered.
To circumvent this limitation, applications may advertise their ability to open
all kinds of files through the "*" filetype.
.B gopen
fails when only the 
.B -a
.I application 
argument but no
.I filename
 is present. Thus, to simply launch an application, please use
.B openapp(1) 
instead of
.B gopen
or 
.B gopen -a
.IR application
.P
.SH SEE ALSO
GNUstep(7), openapp(1)
.P
.SH HISTORY
.B gopen
was written November 2001.
.P
This manual page was first written July 2003.
.P
.SH AUTHORS
.B gopen 
was written by Gregory Casamento <greg_casamento@yahoo.com>.
.P
This man page was written by Martin Brecher <martin@mb-itconsulting.com>.


Tools/gopen.m
/* This tool opens the appropriate application from the command line
   based on what type of file is being accessed. 

   Copyright (C) 2001 Free Software Foundation, Inc.

   Written by:  Gregory Casamento <greg_casamento@yahoo.com>
   Created: November 2001

   This file is part of the GNUstep Project

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License
   as published by the Free Software Foundation; either version 2
   of the License, or (at your option) any later version.
    
   You should have received a copy of the GNU General Public  
   License along with this library; see the file COPYING.LIB.
   If not, write to the Free Software Foundation,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

   */

#include <Foundation/NSArray.h>
#include <Foundation/NSDictionary.h>
#include <Foundation/NSString.h>
#include <Foundation/NSProcessInfo.h>
#include <Foundation/NSTask.h>
#include <Foundation/NSAutoreleasePool.h>
#include <Foundation/NSException.h>
#include <Foundation/NSFileManager.h>
#include <Foundation/NSFileHandle.h>
#include <Foundation/NSPathUtilities.h>
#include <Foundation/NSValue.h>
#include <Foundation/NSUserDefaults.h>
#include <AppKit/NSApplication.h>
#include <AppKit/NSWorkspace.h>

static NSString*
absolutePath(NSFileManager *fm, NSString *path)
{
  path = [path stringByStandardizingPath];
  if ([path isAbsolutePath] == NO)
    {
      path = [[fm currentDirectoryPath] stringByAppendingPathComponent: path];
    }
  return path;
}

int
main(int argc, char** argv, char **env_c)
{
  CREATE_AUTORELEASE_POOL(pool);
  NSEnumerator *argEnumerator = nil;
  NSWorkspace *workspace = nil;
  NSFileManager *fm = nil;
  NSString *arg = nil;
  NSString 
    *editor = nil,
    *terminal = nil;
  NSString 
    *application = nil, 
    *filetoopen = nil,
    *filetoprint = nil,
    *nxhost = nil;

#ifdef GS_PASS_ARGUMENTS
  [NSProcessInfo initializeWithArguments:argv count:argc environment:env_c];
#endif
  argEnumerator = [[[NSProcessInfo processInfo] arguments] objectEnumerator];
  workspace = [NSWorkspace sharedWorkspace];
  fm = [NSFileManager defaultManager];
  
  // Default applications for opening unregistered file types....
  editor = [[NSUserDefaults standardUserDefaults]
    stringForKey: @"GSDefaultEditor"];
  terminal = [[NSUserDefaults standardUserDefaults]
    stringForKey: @"GSDefaultTerminal"];

  // Process options...
  application = [[NSUserDefaults standardUserDefaults] stringForKey: @"a"];
  filetoopen = [[NSUserDefaults standardUserDefaults] stringForKey: @"o"];
  filetoprint = [[NSUserDefaults standardUserDefaults] stringForKey: @"p"];
  nxhost = [[NSUserDefaults standardUserDefaults] stringForKey: @"NXHost"];

  if (application)
    {
      // Don't start the application itself but use it for file opening. 
      //      [workspace launchApplication: application];
    }
  
  if (filetoopen)
    {
      filetoopen = absolutePath(fm, filetoopen);
      [workspace openFile: filetoopen
	  withApplication: application];
    }

  if (filetoprint)
    {
      filetoprint = absolutePath(fm, filetoprint);
      puts("Not implemented");
    }

  if (nxhost)
    {
      puts("Not implemented");
    }

  if (argc == 1)
    {
      NSFileHandle	*fh = [NSFileHandle fileHandleWithStandardInput];
      NSData		*data = [fh readDataToEndOfFile];
      NSString		*tempFile = [NSTemporaryDirectory()
	stringByAppendingPathComponent: @"openfiletmp"];
      NSNumber		*processId = [NSNumber numberWithInt:
	[[NSProcessInfo processInfo] processIdentifier]];

      tempFile = [tempFile stringByAppendingString: [processId stringValue]];
      tempFile = [tempFile stringByAppendingString: @".txt"];
      [data writeToFile: tempFile atomically: YES];
      [workspace openFile: tempFile withApplication: editor];      
    }

  [argEnumerator nextObject]; // skip the first element, which is empty.
  while ((arg = [argEnumerator nextObject]) != nil)
    {
      NSString *ext = [arg pathExtension];
      
      if ([arg isEqualToString: @"-a"])
	{
	  // skip since this is handled above...
	  arg = [argEnumerator nextObject];
	}
      else if ([arg isEqualToString: @"-o"])
	{
	  // skip since this is handled above...
	  arg = [argEnumerator nextObject];
	}
      else if ([arg isEqualToString: @"-p"])
	{
	  // skip since this is handled above...
	  arg = [argEnumerator nextObject];
	}
      else if ([arg isEqualToString: @"-NXHost"])
	{
	  // skip since this is handled above...
	  arg = [argEnumerator nextObject];
	}
      else // no option specified
	{
	  NS_DURING
	    {
	      BOOL isDir = NO, exists = NO;

	      exists = [fm fileExistsAtPath: arg isDirectory: &isDir];
	      if (exists == YES)
		{
		  arg = absolutePath(fm, arg);
		}
	      if (exists && !isDir && [fm isExecutableFileAtPath: arg])
		{
		  [workspace openFile: arg withApplication: terminal];
		}
	      else // no argument specified
		{
		  // First check to see if it's an application
		  if ([ext isEqualToString: @"app"]
		    || [ext isEqualToString: @"debug"]
		    || [ext isEqualToString: @"profile"])
		    {
		      [workspace launchApplication: arg];
		    }
		  else
		    {
		      if (![workspace openFile: arg
			       withApplication: application])
			{
			  // no recognized extension,
			  // run application indicated by environment var.
			  NSLog(@"Opening %@ with %@",arg,editor);
			  [workspace openFile: arg withApplication: editor];
			}
		    }
		}
	    }
	  NS_HANDLER
	    {
	      NSLog(@"Exception while attempting open file %@ - %@: %@",
		    arg, [localException name], [localException reason]);
	    }
	  NS_ENDHANDLER
	}
    }
  RELEASE(pool);
  exit(EXIT_SUCCESS);
}
